yabt/dot.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019 Resonai Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Creates the dot graph
+~~~~~~~~~~~~~~~~~
+
+:author: Dana Shamir
+"""
+
+
+from .caching import get_prebuilt_targets
+from .config import Config
+
+
+TARGETS_COLORS = {'AptPackage': 'brown4',
+                  'CppGTest': 'deepskyblue4',
+                  'CppLib': 'blue',
+                  'CppProg': 'cornflowerblue',
+                  'CustomInstaller': 'brown',
+                  'Proto': 'green',
+                  'Python': 'red',
+                  'PythonPackage': 'purple',
+                  'PythonTest': 'pink'
+                  }
+
+
+def get_not_buildenv_targets(build_context):
```

Comment:
most of the times when I found myself coding some graph algo, I also found that [networkx](https://networkx.github.io/documentation/stable/reference/algorithms/index.html) already had what I wanted (or at least a heavily tested & optimized building block I could use)

Reply:
Even this ? I Incited @dana-resonai to do this - it's a pretty specific case: all the decedents where the heritage line is not through a forbidden set  ? e.g the notorious BGL lets you write shit like this by using predicates, but it's barely readable.

Reply:
could be - there are lots of algos available there, many with names that probably tell you a lot more than they tell me :-)
but even if not, I'm sure there are applicable building blocks (like checking if there's a path between A & B that goes through a certain set of nodes)

Reply:
- I looked at the traversal algorithms, and there aren't any
- "building it from building blocks" is a horrendous idea complexity wise. checking for a path between A & B is graph traversal for itself !